metrics_ci_integration/lib/jenkins/model/jenkins_query_limits.dart

```diff
@@ -0,0 +1,213 @@
+import 'package:meta/meta.dart';
+
+/// A class representing a range specifier for array-type properties in
+/// Jenkins API requests.
+class JenkinsQueryLimits {
+  /// The lower bound of a range specified.
+  final int lower;
+
+  /// The lower bound of a range specified.
+  final int upper;
+
+  /// Creates a range specifier with given [lower] and [upper] bounds.
+  const JenkinsQueryLimits._({this.lower, this.upper});
+
+  /// Checks [number] to be non-`null` positive integer;
+  ///
+  /// Throws [ArgumentError] if the [number] either is `null`
+  /// or [num.isNegative].
+  @visibleForTesting
+  static void checkValue(int number) {
+    if (number == null) {
+      throw ArgumentError.notNull('number');
+    } else if (number.isNegative) {
+      throw ArgumentError.value(number, 'number', 'must be non-negative');
+    }
+  }
+
+  /// Checks [number] to be valid string representing a range-specifier edge.
+  static bool _checkValueString(
+    String number, {
+    bool allowEmpty = true,
+  }) {
+    if (number == null) return false;
+    if (allowEmpty && number.isEmpty) return true;
+    final result = int.tryParse(number);
+    return result != null && result >= 0;
+  }
+
+  /// Checks the range edges [begin] and [end] to be non-`null` positive
+  /// integers and [end] to be greater than or equal to [begin].
+  ///
+  /// Throws [ArgumentError] if:
+  ///   * one of [begin] or [end] is null;
+  ///   * one of [begin] or [end] is negative;
+  ///   * [end] is less than [begin].
+  @visibleForTesting
+  static void checkRange(int begin, int end) {
+    if (begin == null || end == null) {
+      throw ArgumentError('both begin and end must be specified');
+    } else if (begin.isNegative || end.isNegative) {
+      throw ArgumentError('both begin and end must be non-negative');
+    } else if (end < begin) {
+      throw ArgumentError('end must be greater than or equal to begin');
+    }
+  }
+
+  /// Creates an empty range specifier with no edges.
+  ///
+  /// Requests with no range-specifier set will fetch the full data list.
+  const JenkinsQueryLimits.empty() : this._();
+
+  /// Parses a [query] with range-specifier to an instance of [JenkinsQueryLimits].
+  ///
+  /// Throws [FormatException] if parsing fails.
+  factory JenkinsQueryLimits.fromQuery(String query) {
+    if (query.startsWith('{') && query.endsWith('}')) {
+      final limits = query
+          .split(',')
+          .map((substring) => substring.replaceAll(RegExp(r'[\{\}]'), ''))
+          .toList();
+
+      final limitsValid = (limits.length == 1 &&
+              _checkValueString(limits.first, allowEmpty: false)) ||
+          (limits.length < 3 && limits.every(_checkValueString));
```

Comment:
```suggestion
            final limitsValid = limits.length < 3 &&
          limits.every((limit) =>
              _checkValueString(limit, allowEmpty: limits.length != 1));
```
In the current implementation, the limits are valid if the query equals to '{}'. Maybe we can fix it like this and simplify this statement?